### PR TITLE
common,osd: add hash algorithms for dedup fingerprint 

### PR DESCRIPTION
**Pull request by myoungwon**
add sha256, sha512 to use those algorithms as fingerprint IDs for dedup tier

Signed-off-by: Myoungwon Oh <omwmw@sk.com>

### CODE REVIEW

File: src/common/buffer.cc [hunk 1]
```diff
@@ -2055,6 +2057,28 @@ sha1_digest_t buffer::list::sha1()
   return sha1_digest_t(fingerprint);
 }
 
+sha256_digest_t buffer::list::sha256()
```
Comment (tchaikov) [inline src/common/buffer.cc]: i'd suggest implement them as free functions. as they are not tightly coupled with `bufferlist`, instead they are basically the consumers of `list::buffers()` member function.

**Comment by mykaul** [comment]
What's the rational for adding MD5?

**Comment by mattbenjamin** [comment]
@mykaul re: md5:  that's a good question--there is a strong preference in the security community to avoid use of md5, it cannot be used in FIPS 140 certified applications, etc. If the rationale is hashing performance, blake2b has become popular.  From what I understand, in the sha family, sha-512 is actually faster than sha-256, while sha3/keccak are significantly slower.

**Comment by liewegas** [comment]
Yeah, we can drop MD5 (no reason it should ever be used).  And sha2-512 is faster than sha2-256...

File: src/common/buffer.cc (hunk 1 above)
Comment (liewegas) [reply src/common/buffer.cc]: All of the current hashing algorithms are bufferlist methods.  Making them freestanding would be fine, but we can do that as a separate change/cleanup?

**Comment by mykaul** [comment]
I'm wondering if we can use xxhash and the rational/hope is that we can re-use the hash calculated across the stack, instead of each layer recalculating it.

I would not use SHA2-512  over SHA2-256 - I doubt it's faster than 256 and it requires twice the memory to store.

**Comment by mattbenjamin** [comment]
Results showing sha-512 to be faster than sha-256 are widely reported, e.g., https://en.wikipedia.org/wiki/Secure_Hash_Algorithms .  Truncated versions of sha-512 (sha-512/224, sha-512/256) exist.  xxhash is not a cryptographic hash function--that would rule it out for dedup, traditionally, I believe.

**Comment by myoungwon** [comment]
Sorry for late response. 
Yes, we can drop md5. However, my intention was to minimize the size of fingerprint index by using md5. As you know, the size of fingerprint index (chunk a : hash, chunk b : hash ...) is a big problem in terms of deduplication implementation. There are two simple solutions. First is the large chunk and second is using the small length of hash. With regarding to the length of hash, md5 is better than other if users can manage such collision problem depending on the status of the pool.

Anyway, I agree with the security problem of md5. so dropping md5 is also good to me.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-06-09_04:37:10-rados-wip-kefu-testing-2019-06-09-0009-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-06-11_01:51:33-rados-wip-kefu-testing-2019-06-09-0009-distro-basic-smithi/

test failures tracked by 

- http://tracker.ceph.com/issues/17882
- http://tracker.ceph.com/issues/40120
- the misdirected op issue is not related to this issue.

File: src/common/buffer.cc (hunk 1 above)
Comment (tchaikov) [reply src/common/buffer.cc]: #28486

**Pull request merged**